### ISSUE TITLE
Proposal: defer requests to improve testing pending states

### ISSUE THREAD

**Issue by jelhan** [miragejs/discuss#35]
_TL;TR: Providing three additional methods on `server` would make testing pending states a pleasure._

Testing application state while requests are pending is currently not that smooth with mirage as it could be. This proposal tries to address that issues.

Ember's test helper `waitFor` and `waitUntil` provides a nice API for testing pending states but mirage's `server.timing` configuration does not fit in well. To test pending states a timing must be chosen that does not resolve before all assertions for the pending state are done, but on the other hand does not block the tests longer than necessary. This is even more complex cause ember's runloop and timers interacting in a non predictable way. Depending on the device used for running the tests or if the browser tab has focus, a runloop might take way longer than 1ms. So the timing has to be chosen to be long enough for all use cases and you may still see flickering tests if executing in browser (`http://localhost:4200/tests`) and not having focus on the tab.

The only other solution I'm aware of is using `RSVP.defer` and overriding route handlers in each test. But that will produce unreadable and unmaintainable testing code - especially if a route handler is triggered more than once.

I like to propose a new API to mirage to support that use case. It should allow to defer requests and resolve them not before a predictable point of time in your test. To do so `server` object should provide three additional methods: `deferRequests()`, `resolveRequests()` and `undeferRequests()`.

`server.deferRequests()` starts deferring requests. Every request fired afterwards is not settled before `server.resolveRequests()` is called. Deferring requests could be stopped by `server.undeferRequests()`. This will also settle all pending requests.

A test using it could look this:

```
module('Acceptance | post', function(hooks) {
  setupApplicationTest(hooks);
  setupMirage(hooks);

  test('create a new post', async function(assert) {
    await visit('/posts/new');
    await fillIn('[data-test-input=title', 'A new post');
    await fillIn('[data-test-input=body', 'Some text for our new post');

    // defer responses to all requests
    server.deferRequests();

    // do stuff that triggers an API request
    click('[data-test-button=submit]');

    // test application while API requests are pending
    await waitFor('[data-test-button=submit] [data-test-loading-spinner]', {
      timeoutMessage: 'loading spinner has not appeared'
    });
    assert.ok(
      find('[data-test-button=submit]').disabled,
      'submit button is disabled while save request is pending'
    );

    // resolve all pending API requests
    server.resolveRequests();
    await settled();

    // test application after API requests have been settled
    assert.ok(
      find('[data-test-button=submit]').disabled,
      'submit button is enabled again after save request is settled'
    );
  });
});
```

## Overview

```
/**
 * Start deferring requests. Pass optional `path` and `options` arguments to only defer some requests.
 *
 * `path` must match a path of a route registered before.
 * `server.namespace` configuration is taken into consideration.
 * If `path` contains a dynamic segment also it's name must match the one used for route registration.
 *
 * `options` only supports filtering by HTTP request methods. To do so, pass an array of HTTP request methods under
 * `methods` key. Allowed values are: `GET`, `POST`, `PUT`, `PATCH`, `DELETE`, `OPTIONS`, `HEAD`. Additionally
 * `DEL` is supported as an alias for `DELETE`. The value is case-insensitive.
 *
 * `path` is considered a resource helper, if it does not start with a protocol and domain (full URL) or with a slash.
 * In that case it registers deferring for same endpoint, method combination as resource helper does for route
 * registration.
 *
 * @method deferRequests
 * @param  {String} [path]
 * @param  {Object} [options]
 * @return void
 * @public
 */
deferRequests(path, options) {}

/**
 * Cancels deferring requests. Pass optional `path` and `options` arguments to only cancel deferring of some requests.
 * Resolves all deferred requests for affected routes.
 *
 * @method resolveRequests
 * @param  {String}   [path]
 * @param  {Object}   [options]
 * @return {Integer}  number of requests resolved
 * @public
 */
undeferRequests(path, options = {}) {}

/**
 * Resolves deferred requests. Pass optional `path` and `options` arguments to only resolve some deferred requests.
 *
 * @method resolveRequests
 * @param  {String}   [path]
 * @param  {Object}   [options]
 * @return {Integer}  number of requests resolved
 * @public
 */
resolveRequests(path, options = {}) {}
```

## Limit to some requests

All methods should accept a `path` as first argument to only defer, undefer or resolve requests for a specific endpoint. The path must match a registered route. E.g. `server.deferRequests('/posts')` would only defer requests for `/posts` endpoint.

Path is handled the same as on [route registration](http://www.ember-cli-mirage.com/docs/v0.4.x/defining-routes/). It could contain a protocol and domain. If not current `server.namespace` configuration is taken into consideration.

This could be further limited by an `options` hash provided as second argument. The only option supported so far is `methods`. It takes an array of strings. Allowed values are: `'GET'`, `'POST'`, `'PUT'`, `'DELETE'`, `'PATCH'`, `'HEAD'`, `'OPTIONS'`. Additionally `DEL` is supported as an alias for `DELETE` to be inline with verbs used on route registration. The value is case-insensitive. If `methods` option is missing, all requests to specified `path` are deferred independently of method.

Additionally [resource helper](http://www.ember-cli-mirage.com/docs/v0.4.x/shorthands/) are supported. If a path does not start with a protocol or a slash, it's considered a resource helper and will register deferring for the same routes as on route registration.

Another call to `deferRequests()` with the same path will override an earlier one.

Some examples:

```
server.namespace = '';

/*
 * register deferring
 */
server.deferRequests('/companies');
server.deferRequests('/users', { methods: ['GET', 'POST'] });

// this requests will be deferred
jQuery.ajax({ method: 'GET',     url: '/companies' });
jQuery.ajax({ method: 'POST',    url: '/companies' });
jQuery.ajax({ method: 'PUT',     url: '/companies' });
jQuery.ajax({ method: 'PATCH',   url: '/companies' });
jQuery.ajax({ method: 'DELETE',  url: '/companies' });
jQuery.ajax({ method: 'HEAD',    url: '/companies' });
jQuery.ajax({ method: 'OPTIONS', url: '/companies' });

jQuery.ajax({ method: 'GET',     url: '/users' });
jQuery.ajax({ method: 'POST',    url: '/users' });

// this requests won't be deferred (only some examples)
jQuery.ajax({ method: 'GET',     url: '/companies/1' });
jQuery.ajax({ method: 'PATCH',   url: '/users' });
jQuery.ajax({ method: 'DELETE',  url: '/users' });

// using a resource helper
server.deferRequests('customers');
// is just a short for
server.deferRequests('/customers',     { methods: ['GET', 'POST'] });
server.deferRequests('/customers/:id', { methods: ['GET', 'PUT', 'PATCH', 'DELETE'] });

/*
 * resolving requests
 */
server.resolveRequests('/users', { method: 'GET' });
// POST request to /users endpoint is still deferred

// this request is deferred cause deferring of GET requests
// for /users endpoint was not stopped
fetch('/users', { method: 'GET' });

/*
 * cancel deferring of requests
 */
server.undeferRequets('/users');
// new requests to '/users' pendingpoint won't be deferred
// pending POST request to /users endpoint will be resolved
```

## Timing

If route handler returns a promise, requests must not resolve before that one is settled.

```
import { later } from '@ember/runloop';

// route handler which will resolve after 50ms
server.get('/users', () => {
  return new Promise((resolve) => {
    later(() => {
      resolve();
    }, 50);
  });
});

server.deferRequests();
jQuery.ajax({
  method: 'GET',
  url: '/users'
});

// wait 25 ms
await new Promise((resolve) => {
  later(() => {
    resolve();
  }, 25);
});
server.resolveRequests();

// pending promise not fulfilled yet
await settled();
// pending promise fulfilled after 50ms
```

The same applies for timing parameter if used (even so I don't see any use case for that).

## Implementation

An example implementation is provided here: miragejs/ember-cli-mirage#1345 Use it to test out the provided API in real-world scenarios. Please report bugs of the concrete implementation in the PR. This issue should be limited to the overall discussion.

## Open questions

- Did I missed the obvious way solution to the issue with existing API? Hopefully not.  Otherwise this was a fair amout of wasted time...
- Upcoming versions of mirage might drop pretender.js and implement the mocked API using [Service Worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) or [express.js](http://expressjs.com/). Would this new API add any limitations to that ones?
- I'm not totally convinced of naming - especially `undeferRequests()` feels a little bit strange. I also thought of `deferResponses()` instead of `deferRequests()` and `settleRequests()` instead of `resolveRequests()`.

**Comment by samselikoff** [comment]
FYI: Transferred this to our Discuss repo, our new home for more open-ended conversations about Mirage!

If things become more concrete + actionable we can create a tracking issue in the main repo.

**Comment by samselikoff** [comment]
There was some implementation work done in https://github.com/miragejs/ember-cli-mirage/pull/1345

**Comment by ctaylor-nurx** [comment]
Hopping on this thread to ask what would be needed to flesh out/move this feature forward? Is there any consensus on the API? I'd be happy to work on fleshing out test cases in the example implementation or take it in another direction for interface/configuration. 

Really, we're just wanting to expose pretender's underlying ability to manually manage request timing: https://github.com/pretenderjs/pretender#timing-parameter

```
const server = new Pretender(function() {
  // songHandler will only execute once you manually resolve the request
  this.get('/api/songs', songHandler, true);
});

// resolve a request like this
server.resolve(theXMLHttpRequestThatRequestedTheSongsRoute);
```

Should mirage just follow a similar API pattern of deferring a request?

```
// instead of setting the timing
this.get("/api/songs", songHandler, { timing: 4000 })
// we can set defer: true
this.get("/api/songs", songHandler, { defer: true })
// then we can go manually trigger request resolutions
this.deferredRequests.forEach((request) => request.resume());
```

**Comment by samselikoff** [comment]
Probably yes – I'd love to get the lowest-level thing into Mirage so folks can start experimenting with it, and then we can layer on higher-level APIs on later.

What's your use case for this right now?

**Comment by ctaylor-nurx** [comment]
My use case is that I'm writing tests for application loading states for in flight ajax queries. Right now I'm using the `timing` delay, but I have to arbitrarily pick a number big enough to let my test code execute, but short enough to not introduce too much extra test delay. For performance reasons, I'd rather be able to trigger the response rather than guess at a wait for the response.

**Comment by samselikoff** [comment]
Ya agree we need an API to handle this.

The problem with the Pretender API is it would require folks to basically re-write all handlers that they want to pause in their tests.

Think the API in the referenced PR was a decent start but could use some work. Happy to hash it out here but it definitely needs a bit of thought.

Maybe something like

```js
server.deferResponses() // defer all responses
server.deferResponses(({ request, response }) => request.url === 'GET' ) // defer some responses based on request or response data
server.resolveResponses() // resolve all responses. should this be called release? unpause?
server.resolveResponses(({ request, response }) => response.code === 200 ) // same thing
```

(I think it should be `deferResponses` not `deferRequests` right?)

**Comment by samselikoff** [comment]
Also maybe we don't need `server.resolveResponses`, we just expose `server.deferredResponses()` as an array and you can call `.resolve()` on them.

**Comment by ctaylor-nurx** [comment]
You're right. deferResponses is clearer. I was thinking more along the line of deferring a request response by saying deferRequest. deferReponse is more to the point. 

I do wonder, for the general case, if server.pauseResponses(filterFunc) and server.resumeResponses(filterFunc) would be clearer?
Then server.deferredResponses would be a publicly accessible array of { request, response } object with a .resume() method?

So here's an idea for a more fleshed out api idea
```
/*
   * @method pauseResponses
   * @param  {function} filter
   * @param {Integer} timeout
   * @callback filter
   * @param {request} request
   * @param {response} response
*/
pauseResponses(filter, timeout)

/*
   * @method resumeResponses
   * @param  {function} filter
   * @callback filter
   * @param {request} request
   * @param {response} response
*/
resumeResponses(filter)

/*
   * @method getPausedRequests
   * @returns  {array[object]} [pausedRequest]
   * @type pausedRequest
   * @prop {request} pausedRequest.request
   * @prop {response} pausedRequest.response
   * @method pausedRequest.resume()
*/
getPausedRequests()
[{ request, response, resume() }]


// examples
const songFilter = (request, response) => {
  return request.url.includes('songs') || !!response.body.data.songs;
};
server.pauseResponses(songFilter, timeout);

fetch('/api/songs?artist=elvis').then((resp) => console.log('songs', resp));
fetch('/api/lyrics?artist=elvis').then((resp) => console.log('lyrics', resp));
fetch('/api/albums/5/songs?artist=elvis').then((resp) => console.log('album 5 songs', resp));

const pausedRequests = server.getPausedRequests();
// returns an array of { request, response, resume(delay) }
pausedRequests.pop().resume();
// album 5 songs resp

server.resumeResponses(songFilter);
// songs resp
server.resumeResponses();
// lyrics resp
```

**Comment by samselikoff** [comment]
Lookin really good.

Correct me if I'm wrong but the `/lyrics` request would never be paused, right?

Also, I think the signature for the filter function should be an object, just because destructuring makes it just as easy to access, better if just accessing one arg, and also helps future-proof if we want to add more args later.

`server.pauseResponses(({ response }) => response.body === '')`

Also I've used a lot of `getSomething()` methods in Mirage but also started using getters, I think we should use getters going forward. So `server.pausedRequests` could be a property.

The `resume()` method idea is great. Maybe we could bikeshed the name a bit... release? resolve? Also I do like "pause" but maybe "defer" would be better terminology to use given the familiarity with the promise defer spec.

Finally what's the `timeout` arg do?

This feature is going to be f'ing killer!!

**Comment by ctaylor-nurx** [comment]
Thanks for the feedback, can definitely do the destructuring pattern. Was aiming for consistency with the approach from pretender, but with docs, it should be clear enough. I'm happy to use whatever name for resume. thinking pause/resume or defer/release. defer/continue maybe? Can definitely go with implementing a getter. Mostly don't want to allow for accidental mutations, which can be achieved with a private lookup or something else similar. 

For lyrics, you're right. I forgot to add another fetch example. I'll clean that up. 

the timeout is to set a max timeout so that any deferred requests will eventually resolve so as not to freeze the execution indefinitely in case the clearing code fails to execute. More as a backup in case you mess up your future code. similarly, would want to make sure to flush everything when doing a server.shutdown(). 

I can start coding these up in a PR and we can adjust the naming of things more from there.

**Comment by samselikoff** [comment]
Just thinking out loud.. there might be a lower-level API that we should land first. Something like `server.modifyResponse`.

....*several minutes later*

And now I'm thinking a middleware API would address this use case too. I wonder if we should invest our efforts there instead.

```js
let unpause
server.use({ req, res } => {
  if (req.url.includes('songs')) {
    return new Promise(resolve => {
      unpause = resolve
    })
  }
})

// later

unpause()
```

We could layer some sugar on top but this seems like a fundamental missing piece. And theres many other issues that have been asking about a middleware API for a while.

**Comment by ctaylor-nurx** [comment]
So, sidetrack... while reading through the mirage server source and pretender source, I did figure out a way to access/pass through the underlying pretender timer api as well as a way to resolve manual handlers.

```
function resolveManualRequests() {
  server.pretender.requestReferences.forEach((r) => {
    if (server.pretender.requiresManualResolution(r.request.method, r.request.url)) {
      server.pretender.resolve(r.request);
    }
  });
}

server.get('/my-models', { timing: true });
await delay(1000);
resolveManualRequests();
// response returns after 1000ms
```

This makes a lot of what we'd like to accomplish very straightforward in pretender now that I figured it out from the undocumented bits. I'm going to suggest to pretender that `prentender.resolve()` just resolves any requiresManualResolution requests sitting around in the queue. In the meantime, we have a fairly straightforward way to access the functionality and wrap it in a better interface.

**Comment by jelhan** [comment]
Thanks a lot for taking this one up. I planned to follow-up on the feature but haven't found the time yet.

I like the middleware idea. Many people are used to that architecture thanks to express. And it would allow features like this to be implement as plugins to a tiny mirage core.

The API discussion is going in the right direction in my opinion. But there are many open questions left:

- Should the `timing` option have any affect on deferred requests? I don't think so. It's confusing if a request does not resolve immediately if resumed explicitly.
- Do we need a `timeout` option on `Server.deferRequests()`? I think it's less complicated and confusing if deferred requests are only resolved manually. We may emit a warning or throw an error if there are still deferred requests on shutdown.
- What is the exact type of the `request` argument passed to the callback in `Server.deferRequests(({ request }) => true)`? Using `Request` interface of the Fetch API might be a good choice.
- Is the route handler called before a deferred request is resolved? The current proposals assumes that the route handler runs immediately and the response is already available to decide if a request should be deferred or not. This timing would be a relevant part of the public API as it determines if changes to mirage database while a request is deferred affects it's response or not.

I think these questions show the complexity in coming up with a stable API for such a feature. This is a good argument to experiment with these one in a plugin before moving to core.

**Comment by ctaylor-nurx** [comment]
I think what we need in order to implement a proper middleware layer is the idea of "adding" a handler rather than replacing a handler, and differentiating between a middleware step and an ending/returning response step. In express, middleware can call next() to move to the next step, or response.send() to end the request.

```
app.get('/user/:id', (request, response, next) => {
  // do stuff
  next()
})
```

for us, perhaps we extend with a next callback as well? The response in connect is a node http server response: https://nodejs.org/api/http.html#http_class_http_serverresponse

so we'd want to differentiate between response.end() and next() to see if we're middleware or just the handler. If we want to follow node/express's existing API patterns

`.use()` is a generic for all verbs, but here's a more specific example
```
// middleware example with the express pattern
server.get('/user/:id', (schema, request, response, next) => {
  // do stuff
  next();
}, options)

//  end handler/traditional handler with the express pattern
server.get('/user/:id', (schema, request, response, next) => {
  // do stuff
  request.end(schema.users.all());
}, options);
```

**Comment by rreckonerr** [comment]
Hi!
Is there any follow-up to this discussion? Is there any workaround for now?
